book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -165,49 +227,232 @@ each rooted at the highest finalized block. Each chain consists of a map from
 heights to blocks. Chains are stored using an ordered map from difficulty to
 chains, so that the map ordering is the ordering of best to worst chains.
 
-- XXX fill in details on exact types
-
-- **Chain**: `(im::OrdMap<block::Height, Arc<Block>>, HashSet<Nullifier>, HashSet<block::Hash>)`
-  - Ord impl is ordered by work
-  - push => add a block to the end of a chain, does contextual verification
-  checks, extracts info from block for extra data sets
-  - pop => remove the lowest block
-  - fork => create a new chain fork based on a given block(hash) within
-  another chain, calls push repeatedly
-- **ChainSet**: `(BTreeSet<Chain>, BTreeMap<block::Height, Arc<Block>>)`
-  - `fn finalize(&mut self) -> Arc<Block>`
-  - `fn commit_block(&mut self, block: Arc<Block>) -> Result<(), Err>`
-    - iterate through chains, if the parent of `block` is the tip of any
-    chains, try to push block onto that chain
-    - if not, iterate through chains, checking if the parent is contained in
-    that chain
-    - if so, fork the chain at the parent and try to push `block` onto that
-    fork and add newly extended chain if the push succeeds
-    - if not, queue block, prune queued blocks that are below the reorg limit
-
-
-- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
-  spent by some block etc.) to speed up checks.
-
-When a new block extends the best chain past 100 blocks, the old root is
-removed from the non-finalized state and committed to the finalized state.
+### `Chain` Type
+[chain-type]: #chain-type
 
-## Committing non-finalized blocks
+The `Chain` type consists of a set of blocks, containing the non-finalized
+portion of the chain it represents where the lowest height block's parent is
+the tip of the finalized state. All of the other members of `Chain` cache
```

Comment:
The first sentence is long and has a lot of sub-clauses.  I think it would be clearer if it was split up into chunks.  And, rather than jumping into what data it contains, it would be better to open with a description of what it represents:

> The `Chain` type represents a chain of blocks.  Each block represents an incremental state update, and the `Chain` type caches the cumulative state update from its root to its tip.

In the current text, the description of how the `Chain` type relates to the finalized state is mixed with the sentence that describes what it represents.  But if what it represents is described first, then how it fits with other parts of the application can come next:

> The `Chain` type is used to represent the non-finalized portion of a complete chain of blocks rooted at the genesis block.  The parent block of the root of a `Chain` is the tip of the finalized portion of the chain.

Finally, before getting to the data structure definition, it would be helpful to give a summary of the operations that will be defined in this section:

> The `Chain` type supports several operations to manipulate chains ...